types/transducers-js/index.d.ts

```diff
@@ -39,8 +39,18 @@ export function isReduced(x: any): boolean;
 /**
  * Function composition. Take N function and return their composition.
  */
-// Infers correct return type when all arguments have same type.
-export function comp<T extends (x: any) => any>(...args: T[]): T;
+// 2-4 Transducers
+export function comp<A, B, C>(a: Transducer<A, B>, b: Transducer<B, C>): Transducer<A, C>;
+export function comp<A, B, C, D>(a: Transducer<A, B>, b: Transducer<B, C>, c: Transducer<C, D>): Transducer<A, D>;
+export function comp<A, B, C, D, E>(a: Transducer<A, B>, b: Transducer<B, C>, c: Transducer<C, D>, d: Transducer<D, E>): Transducer<A, E>;
+// N identical Transducers
+export function comp<A>(...args: Array<Transducer<A, A>>): Transducer<A, A>;
+// 2-4 arbitrary functions
+export function comp<A, B, C>(a: (a: A) => B, b: (b: B) => C): (a: A) => C;
+export function comp<A, B, C, D>(a: (a: A) => B, b: (b: B) => C, c: (c: C) => D): (a: A) => D;
+export function comp<A, B, C, D, E>(a: (a: A) => B, b: (b: B) => C, c: (c: C) => D, d: (d: D) => E): (a: A) => E;
```

Comment:
Aren't these non-transducer overloads backwards? `comp(f, g)(x)` is `f(g(x))`, meaning the output of `g` is the input of `f`.

Reply:
Yes... wouldn't the transducers overloads be backwards too then?

Reply:
Actually no. The definition of `Transducer` is
```ts
export type Transducer<TInput, TOutput> =
    <TResult, TCompleteResult>(
        xf: CompletingTransformer<TResult, TCompleteResult, TOutput>
    ) => CompletingTransformer<TResult, TCompleteResult, TInput>;
```
which stripping down to the relevant parts is
```ts
type Transducer<TInput, TOutput> = (xf: Thingy<TOutput>) => Thingy<TInput>
```
so the transducer type parameters are actually the opposite direction that they appear in the function.